api/src/server/ext.rs

```diff
@@ -0,0 +1,115 @@
+use crate::{
+    rpc::{ApiError, ApiResult, Request, Response, ResponseObject},
+    server::Context,
+};
+
+use axum::{
+    body::Body,
+    extract::{Extension, Json},
+    response::{IntoResponse, Response as AxumResponse},
+    routing::{post, Router},
+};
+use futures::Future;
+use http::StatusCode;
+use serde::{de::DeserializeOwned, Serialize};
+
+/// Marker trait to ensure handlers are in a good shape.
+pub trait Method<Req: Request, F: Future<Output = ApiResult<Req::Res>>> {
+    fn invoke(self, ctx: Context, req: Req) -> F;
+}
+
+impl<Req, Func, Fut> Method<Req, Fut> for Func
+where
+    Req: Request,
+    Fut: Future<Output = ApiResult<Req::Res>>,
+    Func: FnOnce(Req, Context) -> Fut,
+{
+    fn invoke(self, ctx: Context, req: Req) -> Fut {
+        self(req, ctx)
+    }
+}
+
+pub trait RouterExt {
+    #[must_use]
+    fn mount<M, Req, Fut>(self, method: M) -> Self
+    where
+        M: Method<Req, Fut> + Send + Clone + 'static,
+        Fut: Future<Output = ApiResult<Req::Res>> + Send,
+        Req: DeserializeOwned + Request + Send + 'static,
+        Req::Res: Serialize;
+}
+
+impl RouterExt for Router<Body> {
+    fn mount<M, R, F>(self, method: M) -> Self
+    where
+        M: Method<R, F> + Send + Clone + 'static,
+        F: Future<Output = ApiResult<R::Res>> + Send,
+        R: DeserializeOwned + Request + Send + 'static,
+        R::Res: Serialize,
+    {
+        let handler = move |Json(req): Json<R>, Extension(ctx): Extension<Context>| async {
+            match method.invoke(ctx, req).await {
+                Ok(res) => res.packed().into_response(),
+                Err(e) => e.packed().into_response(),
+            }
+        };
+
+        self.route(&("/v1/".to_owned() + R::METHOD), post(handler))
```

Comment:
What about lift v1 up a layer?

So we'll have something like `router.nest("/v1", api_routes)`